Error when running plugin

Hey,

I took you plugin configuration from quick start and run the `mvn compile` phase. Sadly I'm getting and error: `Cannot invoke "java.lang.Class.isAssignableFrom(java.lang.Class)" because "superClass" is null`

```
Caused by: java.lang.NullPointerException: Cannot invoke "java.lang.Class.isAssignableFrom(java.lang.Class)" because "superClass" is null
    at io.github.kbuntrock.model.DataObject.getContextualType (DataObject.java:346)
    at io.github.kbuntrock.TagLibrary.inspectObject (TagLibrary.java:113)
    at io.github.kbuntrock.TagLibrary.exploreDataObject (TagLibrary.java:77)
    at io.github.kbuntrock.TagLibrary.exploreDataObject (TagLibrary.java:84)
    at io.github.kbuntrock.TagLibrary.inspectObject (TagLibrary.java:114)
    at io.github.kbuntrock.TagLibrary.exploreDataObject (TagLibrary.java:77)
    at io.github.kbuntrock.TagLibrary.exploreDataObject (TagLibrary.java:84)
    at io.github.kbuntrock.TagLibrary.exploreTagObjects (TagLibrary.java:60)
    at io.github.kbuntrock.TagLibrary.addTag (TagLibrary.java:39)
```

Class that has implementation for endpoint:

```java
@Slf4j
@RestController
@RequiredArgsConstructor
public class Foo implements Boo {

  @Override
  public ResponseEntity<Collection<AppInfo>> getDeployedApplications(ILoginName loginName) {
    // ...
  }

}
```

Interface:

```java
@RequestMapping(API_PREFIX)
public interface Boo {

  String API_PREFIX = "/api";

  // @RequestBody ILoginName loginName
  @PreAuthorize("isAuthenticated()")
  @PostMapping(path = "/available-applications")
  ResponseEntity<Collection<AppInfo>> getDeployedApplications(
      @RequestBody ILoginName loginName);

}
```

--- a/openapi-maven-plugin/src/main/java/io/github/kbuntrock/JavaClassAnalyser.java
+++ b/openapi-maven-plugin/src/main/java/io/github/kbuntrock/JavaClassAnalyser.java
@@ -11,7 +11,6 @@ import java.lang.reflect.Parameter;
 import java.util.ArrayList;
 import java.util.List;
 import java.util.Optional;
-import java.util.regex.Matcher;
 import java.util.regex.Pattern;
 import org.apache.commons.lang3.StringUtils;
 import org.apache.commons.lang3.tuple.Pair;
@@ -71,17 +70,14 @@ public class JavaClassAnalyser {
 	private static String createTypeIdentifier(final String typeName) {
 
 		String returnTypeName = typeName;
-		final Matcher matcher = BEGINNING.matcher(returnTypeName);
-		if(matcher.find() && matcher.group().contains(".")) {
-			returnTypeName = returnTypeName.replace(matcher.group(), "");
+		final String[] toReplace = typeName.split("<|>|,");
+		final List<Pair<String, String>> replacementList = new ArrayList<>();
+		for(final String s : toReplace) {
+			final String[] replacementArray = s.split("\\.");
+			replacementList.add(Pair.of(s, replacementArray[replacementArray.length - 1]));
 		}
-		final Matcher matcher2 = FIRST_GENERIC.matcher(returnTypeName);
-		if(matcher2.find() && matcher2.group().contains(".")) {
-			returnTypeName = returnTypeName.replace(matcher2.group(), "<");
-		}
-		final Matcher matcher3 = OTHER_GENERIC.matcher(returnTypeName);
-		if(matcher3.find() && matcher3.group().contains(".")) {
-			returnTypeName = returnTypeName.replace(matcher3.group(), ",");
+		for(final Pair<String, String> pair : replacementList) {
+			returnTypeName = returnTypeName.replace(pair.getLeft(), pair.getRight());
 		}
 
 		return returnTypeName;

--- a/openapi-maven-plugin/src/main/java/io/github/kbuntrock/model/DataObject.java
+++ b/openapi-maven-plugin/src/main/java/io/github/kbuntrock/model/DataObject.java
@@ -5,11 +5,9 @@ import io.github.kbuntrock.configuration.EnumConfigHolder;
 import io.github.kbuntrock.reflection.GenericArrayTypeImpl;
 import io.github.kbuntrock.reflection.ParameterizedTypeImpl;
 import io.github.kbuntrock.reflection.ReflectionsUtils;
-import io.github.kbuntrock.utils.Logger;
 import io.github.kbuntrock.utils.OpenApiDataType;
 import java.lang.reflect.Field;
 import java.lang.reflect.GenericArrayType;
-import java.lang.reflect.GenericDeclaration;
 import java.lang.reflect.ParameterizedType;
 import java.lang.reflect.Type;
 import java.lang.reflect.TypeVariable;
@@ -339,31 +337,7 @@ public class DataObject {
 			} else if(genericType instanceof TypeVariable) {
 
 				// We are in presence of a generic type variable not coming from the outside. Might be coming from generic typing at a parent level.
-				final TypeVariable typeVariable = (TypeVariable) genericType;
-				GenericDeclaration genericDeclaration = typeVariable.getGenericDeclaration();
-				Class superClass = this.getJavaClass().getSuperclass();
-				Type genericSuperClass = this.getJavaClass().getGenericSuperclass();
-				while(!genericDeclaration.equals(superClass) && !superClass.isAssignableFrom(Object.class)) {
-					genericDeclaration = typeVariable.getGenericDeclaration();
-					superClass = superClass.getSuperclass();
-					genericSuperClass = superClass.getGenericSuperclass();
-				}
-				superClass.getDeclaredConstructors();
-				if(genericDeclaration.equals(typeVariable.getGenericDeclaration())) {
-					int i;
-					for(i = 0; i < superClass.getTypeParameters().length; i++) {
-						if(typeVariable.getName().equals(superClass.getTypeParameters()[i].getName())) {
-							if(genericSuperClass instanceof ParameterizedType) {
-								return ((ParameterizedType) genericSuperClass).getActualTypeArguments()[i];
-							}
-						}
-					}
-					// If no substitution has been found yet, the generic type could have been set by a contructor
-					return TypeToken.of(genericType).resolveType(javaType).getType();
-
-				} else {
-					Logger.INSTANCE.getLogger().warn("No generic substitution found for object " + this.getJavaClass().getName());
-				}
+				return TypeToken.of(this.getJavaClass()).resolveType(genericType).getType();
 			}
 
 		}

--- a/openapi-maven-plugin/src/main/java/io/github/kbuntrock/yaml/model/Schema.java
+++ b/openapi-maven-plugin/src/main/java/io/github/kbuntrock/yaml/model/Schema.java
@@ -11,6 +11,7 @@ import com.fasterxml.jackson.annotation.JsonProperty;
 import io.github.kbuntrock.JavaClassAnalyser;
 import io.github.kbuntrock.TagLibraryHolder;
 import io.github.kbuntrock.javadoc.ClassDocumentation;
+import io.github.kbuntrock.javadoc.ClassDocumentation.EnhancementType;
 import io.github.kbuntrock.javadoc.JavadocMap;
 import io.github.kbuntrock.javadoc.JavadocWrapper;
 import io.github.kbuntrock.model.DataObject;
@@ -91,7 +92,7 @@ public class Schema {
 		if(JavadocMap.INSTANCE.isPresent()) {
 			classDocumentation = JavadocMap.INSTANCE.getJavadocMap().get(dataObject.getJavaClass().getCanonicalName());
 			if(classDocumentation != null) {
-				classDocumentation.inheritanceEnhancement(dataObject.getJavaClass(), ClassDocumentation.EnhancementType.FIELDS);
+				classDocumentation.inheritanceEnhancement(dataObject.getJavaClass(), EnhancementType.BOTH);
 			}
 			if(classDocumentation != null && mainReference) {
 				final Optional<String> optionalDescription = classDocumentation.getDescription();

--- a/openapi-maven-plugin/src/test/java/io/github/kbuntrock/SpringClassAnalyserTest.java
+++ b/openapi-maven-plugin/src/test/java/io/github/kbuntrock/SpringClassAnalyserTest.java
@@ -36,6 +36,7 @@ import io.github.kbuntrock.resources.endpoint.generic.GenericityTestTen;
 import io.github.kbuntrock.resources.endpoint.generic.GenericityTestThree;
 import io.github.kbuntrock.resources.endpoint.generic.GenericityTestTwelve;
 import io.github.kbuntrock.resources.endpoint.generic.GenericityTestTwo;
+import io.github.kbuntrock.resources.endpoint.generic.Issue89;
 import io.github.kbuntrock.resources.endpoint.ignore.JsonIgnoreController;
 import io.github.kbuntrock.resources.endpoint.interfacedto.InterfaceController;
 import io.github.kbuntrock.resources.endpoint.map.MapController;
@@ -212,6 +213,19 @@ public class SpringClassAnalyserTest extends AbstractTest {
 
 		final List<File> generated = mojo.documentProject();
 		checkGenerationResult("ut/SpringClassAnalyserTest/genericity_in_super_constructor.yml", generated.get(0));
+	}
+
+	@Test
+	public void issue_89() throws MojoFailureException, IOException, MojoExecutionException {
+
+		final DocumentationMojo mojo = createBasicMojo(Issue89.class.getCanonicalName());
+		final JavadocConfiguration javadocConfig = new JavadocConfiguration();
+		javadocConfig.setScanLocations(Arrays.asList("src/test/java/io/github/kbuntrock/resources/endpoint/generic",
+			"src/test/java/io/github/kbuntrock/resources/dto/genericity/issue89"));
+		mojo.setJavadocConfiguration(javadocConfig);
+
+		final List<File> generated = mojo.documentProject();
+		checkGenerationResult("ut/SpringClassAnalyserTest/issue_89.yml", generated.get(0));
 	}
 
 	@Test
@@ -592,6 +606,7 @@ public class SpringClassAnalyserTest extends AbstractTest {
 		blackList.add(".*\\.GenericityTestS.*$");
 		blackList.add(".*\\.GenericityTestE.*$");
 		blackList.add(".*\\.GenericityTestN.*$");
+		blackList.add(".*\\.Issue89.*$");
 		blackList.add(".*\\.ActionResource");
 		blackList.add(".*\\.GenericDataController");
 		mojo.getApiConfiguration().setBlackList(blackList);

--- a/openapi-maven-plugin/src/test/java/io/github/kbuntrock/resources/dto/genericity/issue89/Bar.java
+++ b/openapi-maven-plugin/src/test/java/io/github/kbuntrock/resources/dto/genericity/issue89/Bar.java
@@ -1,0 +1,12 @@
+package io.github.kbuntrock.resources.dto.genericity.issue89;
+
+/**
+ * This is the Bar object
+ */
+public class Bar extends Boo {
+
+	/**
+	 * the bar string
+	 */
+	private String bar;
+}

--- a/openapi-maven-plugin/src/test/java/io/github/kbuntrock/resources/dto/genericity/issue89/Boo.java
+++ b/openapi-maven-plugin/src/test/java/io/github/kbuntrock/resources/dto/genericity/issue89/Boo.java
@@ -1,0 +1,12 @@
+package io.github.kbuntrock.resources.dto.genericity.issue89;
+
+/**
+ * This is the Boo object
+ */
+public class Boo extends Foo {
+
+	/**
+	 * the boo boolean
+	 */
+	private boolean boo;
+}

--- a/openapi-maven-plugin/src/test/java/io/github/kbuntrock/resources/dto/genericity/issue89/Foo.java
+++ b/openapi-maven-plugin/src/test/java/io/github/kbuntrock/resources/dto/genericity/issue89/Foo.java
@@ -1,0 +1,13 @@
+package io.github.kbuntrock.resources.dto.genericity.issue89;
+
+/**
+ * This is the Foo object
+ */
+public class Foo extends Pair<Foo, Boo> {
+
+	/**
+	 * the foo integer
+	 */
+	private Integer foo;
+
+}

--- a/openapi-maven-plugin/src/test/java/io/github/kbuntrock/resources/dto/genericity/issue89/IBar.java
+++ b/openapi-maven-plugin/src/test/java/io/github/kbuntrock/resources/dto/genericity/issue89/IBar.java
@@ -1,0 +1,14 @@
+package io.github.kbuntrock.resources.dto.genericity.issue89;
+
+/**
+ * This is the IBar object
+ */
+public interface IBar extends IBoo, IX {
+
+	/**
+	 * The bar string
+	 *
+	 * @return the bar string
+	 */
+	String getBar();
+}

--- a/openapi-maven-plugin/src/test/java/io/github/kbuntrock/resources/dto/genericity/issue89/IBoo.java
+++ b/openapi-maven-plugin/src/test/java/io/github/kbuntrock/resources/dto/genericity/issue89/IBoo.java
@@ -1,0 +1,14 @@
+package io.github.kbuntrock.resources.dto.genericity.issue89;
+
+/**
+ * This is the IBoo object
+ */
+public interface IBoo extends IFoo {
+
+	/**
+	 * The boo boolean
+	 *
+	 * @return the boo boolean
+	 */
+	boolean getBoo();
+}

--- a/openapi-maven-plugin/src/test/java/io/github/kbuntrock/resources/dto/genericity/issue89/IFoo.java
+++ b/openapi-maven-plugin/src/test/java/io/github/kbuntrock/resources/dto/genericity/issue89/IFoo.java
@@ -1,0 +1,15 @@
+package io.github.kbuntrock.resources.dto.genericity.issue89;
+
+/**
+ * The IFoo object
+ */
+public interface IFoo {
+
+	/**
+	 * The foo integer
+	 *
+	 * @return the foo integer
+	 */
+	Integer getFoo();
+
+}

--- a/openapi-maven-plugin/src/test/java/io/github/kbuntrock/resources/dto/genericity/issue89/IPair.java
+++ b/openapi-maven-plugin/src/test/java/io/github/kbuntrock/resources/dto/genericity/issue89/IPair.java
@@ -1,0 +1,14 @@
+package io.github.kbuntrock.resources.dto.genericity.issue89;
+
+/**
+ * The IPair object
+ */
+public interface IPair<A, B> extends ISingle<A> {
+
+	/**
+	 * Get the B attribute
+	 *
+	 * @return the B attribute
+	 */
+	B getB();
+}

--- a/openapi-maven-plugin/src/test/java/io/github/kbuntrock/resources/dto/genericity/issue89/ISingle.java
+++ b/openapi-maven-plugin/src/test/java/io/github/kbuntrock/resources/dto/genericity/issue89/ISingle.java
@@ -1,0 +1,14 @@
+package io.github.kbuntrock.resources.dto.genericity.issue89;
+
+/**
+ * The ISingle object
+ */
+public interface ISingle<A> {
+
+	/**
+	 * get the A attribute
+	 *
+	 * @return the A attribute
+	 */
+	A getA();
+}

--- a/openapi-maven-plugin/src/test/java/io/github/kbuntrock/resources/dto/genericity/issue89/IX.java
+++ b/openapi-maven-plugin/src/test/java/io/github/kbuntrock/resources/dto/genericity/issue89/IX.java
@@ -1,0 +1,8 @@
+package io.github.kbuntrock.resources.dto.genericity.issue89;
+
+/**
+ * The IX object
+ */
+public interface IX extends IPair<IFoo, IBoo> {
+
+}

--- a/openapi-maven-plugin/src/test/java/io/github/kbuntrock/resources/dto/genericity/issue89/Pair.java
+++ b/openapi-maven-plugin/src/test/java/io/github/kbuntrock/resources/dto/genericity/issue89/Pair.java
@@ -1,0 +1,12 @@
+package io.github.kbuntrock.resources.dto.genericity.issue89;
+
+/**
+ * This is the Pair object
+ */
+public class Pair<A, B> extends Single<A> {
+
+	/**
+	 * the b attribute
+	 */
+	private B b;
+}

--- a/openapi-maven-plugin/src/test/java/io/github/kbuntrock/resources/dto/genericity/issue89/Single.java
+++ b/openapi-maven-plugin/src/test/java/io/github/kbuntrock/resources/dto/genericity/issue89/Single.java
@@ -1,0 +1,12 @@
+package io.github.kbuntrock.resources.dto.genericity.issue89;
+
+/**
+ * This is the Single object
+ */
+public class Single<A> {
+
+	/**
+	 * the a attribute
+	 */
+	private A a;
+}

--- a/openapi-maven-plugin/src/test/java/io/github/kbuntrock/resources/endpoint/generic/Issue89.java
+++ b/openapi-maven-plugin/src/test/java/io/github/kbuntrock/resources/endpoint/generic/Issue89.java
@@ -1,0 +1,29 @@
+package io.github.kbuntrock.resources.endpoint.generic;
+
+import io.github.kbuntrock.resources.dto.genericity.issue89.Bar;
+import io.github.kbuntrock.resources.dto.genericity.issue89.IBar;
+import java.util.Collection;
+import org.springframework.http.ResponseEntity;
+import org.springframework.web.bind.annotation.PostMapping;
+import org.springframework.web.bind.annotation.RequestBody;
+import org.springframework.web.bind.annotation.RequestMapping;
+
+/**
+ * Webservices about the issue number 89
+ *
+ * @author Kevin Buntrock
+ */
+@RequestMapping(Issue89.API_PREFIX)
+public interface Issue89 {
+
+	String API_PREFIX = "/issue89";
+
+	/**
+	 * This api test the issue n°89
+	 *
+	 * @param myParam my awesome parameter
+	 * @return my also awesome response
+	 */
+	@PostMapping(path = "/youAreAwesome-applications")
+	ResponseEntity<Collection<IBar>> youAreAwesome(@RequestBody Bar myParam);
+}

--- a/openapi-maven-plugin/src/test/resources/ut/SpringClassAnalyserTest/genericity_in_super_constructor.yml
+++ b/openapi-maven-plugin/src/test/resources/ut/SpringClassAnalyserTest/genericity_in_super_constructor.yml
@@ -41,8 +41,19 @@ components:
         myString:
           type: string
         self:
-          $ref: '#/components/schemas/SearchCriteriaChildV6'
+          $ref: '#/components/schemas/SearchCriteriaV6'
         someBoolean:
           type: boolean
         anotherThing:
+          type: boolean
+    SearchCriteriaV6:
+      type: object
+      properties:
+        dateCriteria:
+          $ref: '#/components/schemas/CriteriaDateType'
+        myString:
+          type: string
+        self:
+          $ref: '#/components/schemas/SearchCriteriaV6'
+        someBoolean:
           type: boolean

--- a/openapi-maven-plugin/src/test/resources/ut/SpringClassAnalyserTest/interface.yml
+++ b/openapi-maven-plugin/src/test/resources/ut/SpringClassAnalyserTest/interface.yml
@@ -14,9 +14,10 @@ paths:
       tags:
         - InterfaceController
       operationId: getInterfaceDto
+      description: This endpoint returns an interface
       responses:
         200:
-          description: successful operation
+          description: the returned interface
           content:
             '*/*':
               schema:
@@ -37,6 +38,7 @@ paths:
                         $ref: '#/components/schemas/Authority'
                     uniqueItems: true
                   parentId:
+                    description: Get the parent id
                     type: integer
                     format: int32
                   returned3:

--- a/openapi-maven-plugin/src/test/resources/ut/SpringClassAnalyserTest/issue_89.yml
+++ b/openapi-maven-plugin/src/test/resources/ut/SpringClassAnalyserTest/issue_89.yml
@@ -1,0 +1,124 @@
+---
+openapi: 3.0.3
+info:
+  title: My Project
+  version: 10.5.36
+servers:
+  - url: ""
+tags:
+  - name: Issue89
+    description: Webservices about the issue number 89
+paths:
+  /issue89/youAreAwesome-applications:
+    post:
+      tags:
+        - Issue89
+      operationId: youAreAwesome
+      description: This api test the issue n°89
+      requestBody:
+        content:
+          '*/*':
+            schema:
+              $ref: '#/components/schemas/Bar'
+        description: my awesome parameter
+      responses:
+        200:
+          description: my also awesome response
+          content:
+            '*/*':
+              schema:
+                type: array
+                items:
+                  $ref: '#/components/schemas/IBar'
+components:
+  schemas:
+    Bar:
+      description: This is the Bar object
+      type: object
+      properties:
+        a:
+          description: the a attribute
+          $ref: '#/components/schemas/Foo'
+        b:
+          description: the b attribute
+          $ref: '#/components/schemas/Boo'
+        foo:
+          description: the foo integer
+          type: integer
+          format: int32
+        boo:
+          description: the boo boolean
+          type: boolean
+        bar:
+          description: the bar string
+          type: string
+    Boo:
+      description: This is the Boo object
+      type: object
+      properties:
+        a:
+          description: the a attribute
+          $ref: '#/components/schemas/Foo'
+        b:
+          description: the b attribute
+          $ref: '#/components/schemas/Boo'
+        foo:
+          description: the foo integer
+          type: integer
+          format: int32
+        boo:
+          description: the boo boolean
+          type: boolean
+    Foo:
+      description: This is the Foo object
+      type: object
+      properties:
+        a:
+          description: the a attribute
+          $ref: '#/components/schemas/Foo'
+        b:
+          description: the b attribute
+          $ref: '#/components/schemas/Boo'
+        foo:
+          description: the foo integer
+          type: integer
+          format: int32
+    IBar:
+      description: This is the IBar object
+      type: object
+      properties:
+        a:
+          description: get the A attribute
+          $ref: '#/components/schemas/IFoo'
+        b:
+          description: Get the B attribute
+          $ref: '#/components/schemas/IBoo'
+        bar:
+          description: The bar string
+          type: string
+        boo:
+          description: The boo boolean
+          type: boolean
+        foo:
+          description: The foo integer
+          type: integer
+          format: int32
+    IBoo:
+      description: This is the IBoo object
+      type: object
+      properties:
+        boo:
+          description: The boo boolean
+          type: boolean
+        foo:
+          description: The foo integer
+          type: integer
+          format: int32
+    IFoo:
+      description: The IFoo object
+      type: object
+      properties:
+        foo:
+          description: The foo integer
+          type: integer
+          format: int32

